pkg/operator/controller/ingress/controller.go

```diff
@@ -688,6 +687,23 @@ func (r *reconciler) ensureIngressDeleted(ingress *operatorv1.IngressController)
 			errs = append(errs, fmt.Errorf("failed to get deployment for ingress %s/%s: %v", ingress.Namespace, ingress.Name, err))
 		} else if haveDepl {
 			errs = append(errs, fmt.Errorf("deployment still exists for ingress %s/%s", ingress.Namespace, ingress.Name))
+		} else {
+			// Wait for all the router pods to be deleted. This is important because the router deployment
+			// gets deleted a handful of milliseconds before the router pods process the graceful shutdown. This causes
+			// a race condition in which we clear route status, then the router pod will race to re-admit the status in
+			// these few milliseconds it before it processes the graceful shutdown. The only way to avoid is to wait
+			// until all router pods are deleted.
+			if allDeleted, err := r.allRouterPodsDeleted(ingress); err != nil {
+				errs = append(errs, err)
+			} else if allDeleted {
+				// Deployment has been deleted and there are no more pods left.
+				// Clear all routes status for this ingress controller.
+				if statusErrs := r.clearAllRoutesStatusForIngressController(ingress.ObjectMeta.Name); len(statusErrs) > 0 {
+					errs = append(errs, statusErrs...)
```

Comment:
A concern I have, like the above, which applies to all thing of this pattern is: if you return an error on `ensureIngressDeletion`, it doesn't retry reconcilation.

So in my testing, I found a bug in which it failed the 1st time this was run. It'd then take another 6 minutes to Reconcile again. Why would we not try again if we failed to delete the ingress controller? The user's delete command is being blocked and not sure if this is the right approach (that is, to not even try again). We are stuck in terminating.

Reply:
Side note: The reason it fails on the first time is because `r.client` lists duplicate canary routes for some reason. [Thread here](https://coreos.slack.com/archives/CBWMXQJKD/p1649704358332479)

Reply:
Update: Miciah said it really should reconcile on an error and maybe I was seeing the exponential backoff logic delays and getting confused.

Update2: We found an issue with `r.cache` in which it has duplicate objects. That is being fixed separately.